### PR TITLE
Ports FTL's space manager

### PR DESCRIPTION
**Pull request by Cyberboss**
:cl:
experiment: Greatly improved zlevel and space management
refactor: No maps are preloaded at round start
refactor: Station Z moved to Z3+
refactor: Mining Z moved to Z2
/:cl:

Why: Our current system is a patchwork mess and needs modernization and modularization. This gives us methods to efficiently load, allocate, and unload space

TODO:

- [ ] Finish stitching together SSmapping
- [ ] Switch usage of old defines to use the helper procs
- [ ] Fix other compile errors
- [ ] Check space transitions haven't broken
- [ ] Make map config jsons use the trait system
- [ ] Make transit Z use the new system
- [ ] Make Reebe use the new system
- [ ] Make space ruin placement use the new system
- [ ] Make away missions use the new system
- [ ] Remove leftover z crap

### CODE REVIEW

**Comment by KorPhaeron** [comment]
refactor: Mining Z moved to Z2
refactor: Station Z moved to Z3+

Wont this make people in mining fall into the station via chasm?

**Comment by ike709** [comment]
One less thing for FTL13 to maintain.

**Comment by Cyberboss** [comment]
@KorPhaeron we can just add a z trait IS_ABOVE_NEXT_Z or something

**Comment by Cyberboss** [comment]
The issue with the current arrangement is we're putting the static level (Mining) after the dynamic level (Station) which causes pains

**Comment by vuonojenmustaturska** [comment]
Put a 3x3 zone of lava and death in Z3, move station to Z4

**Comment by nfreader** [comment]
>refactor: Station Z moved to Z3+
>refactor: Mining Z moved to Z2

Wee woo wee woo wee woo oh no you set off the database integrity alarm!

**Comment by Qustinnus** [comment]
Cyber I think the idea is that we'll also have multiZ mining though so won't that be an issue?

**Comment by Cyberboss** [comment]
Well then one of them will have to be offset by the other. Probably mining
= station + n

**Comment by KingPhilipIII** [comment]
I don't understand coding at all but will this make the game implode if we're not careful?

**Comment by AuroranAI** [comment]
Coding clownshoes will make the game implode if you're not careful

**Comment by francinum** [comment]
@KorPhaeron this'd make people fall into centcom, not the station.

**Comment by DoctoreMad** [comment]
Here comes the shitcode train!
Choo choo!

Did you know that on FTL, the map(including lavaland) is generated EVERY TIME you move? Like, every single part of the map, so every FTL jump it generates mining and a docking station and all that fancy stuff, _EVEN IF THERE ISN'T ONE ON THE SECTOR THEY'RE IN?_
AAAAAAAAAAAAA!

**Comment by Cyberboss** [comment]
I'm personally auditing it so that's not gonna happen

**Comment by DoctoreMad** [comment]
Well I mean that thing I mentioned isn't going to happen either way since your PR is unrelated.
Just, man the memories.

**Comment by DoctoreMad** [comment]
>Since your PR is unrelated

I mean it is but
Fuck it I'll edit the above post and delete this one

**Comment by Crazylemon64** [comment]
Oh hey, you ported https://github.com/ParadiseSS13/Paradise/pull/5254 this, effectively

I'll be keeping an eye on this, then

**Comment by Crazylemon64** [comment]
One thing you WILL have to do, is go through the code and decouple hard-coded z-level checks from directly checking numbers - I did that in https://github.com/ParadiseSS13/Paradise/pull/5268

**Comment by Crazylemon64** [comment]
If you wanna do arbitrary amounts of ruins levels, you can just do something like what I have here:
https://github.com/ParadiseSS13/Paradise/blob/master/code/controllers/master_controller.dm#L61-L67

**Comment by Crazylemon64** [comment]
@KorPhaeron You might want to make chasms take a "level ID" instead of a z coordinate, so that the chasms determine which level to fall to based on the ID you give them instead of a hard-coded number

**Comment by Qustinnus** [comment]
doctoremad blame monster for that. That guy wouldnt be able to know what optimizing was even if it slapped him in the face at 200mph

**Comment by Crazylemon64** [comment]
Ah, speaking of optimization, the "space chunk allocator" currently uses a very naive quad-tree implementation that currently does not re-balance. You almost certainly want to re-write the allocation algorithm to something more efficient and scalable

File: code/modules/mapping/space_management/heap_space_level.dm [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	qdel(top, TRUE)
+	return ..()
+
+// Returns whether this zlevel has room for the given amount of space
+/datum/space_level/heap/proc/request(width, height)
+	return top.can_fit_space(width, height)
+
+// Returns a space chunk datum for some nerd to work with - tells them what's safe to write into, and such
+// Call request() first to sanity check that the space CAN be allocated
+/datum/space_level/heap/proc/allocate(width, height)
```
Comment (Crazylemon64) [inline code/modules/mapping/space_management/heap_space_level.dm]: This proc is the one I was talking about - you'll want to rewrite this to something more efficient if you care at all about performance, as of right now it uses a naive imbalanced quad-tree

Comment (Crazylemon64) [inline code/modules/mapping/space_management/heap_space_level.dm]: This is the proc I was talking about - as it stands, it uses a naive quad-tree that does not rebalance

File: code/modules/mapping/space_management/space_level.dm [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+/datum/space_level/proc/apply_transition(turf/open/space/S)
+	if(SSmapping.unbuilt_space_transitions[src])
+		return // Let SSmapping handle this one
+	switch(linkage)
+		if(UNAFFECTED)
+			S.remove_transitions()
+		if(SELFLOOPING,CROSSLINKED)
+			var/datum/space_level/E = get_connection()
+			if(S in transit_north)
+				E = get_connection("[NORTH]")
```
Comment (Crazylemon64) [inline code/modules/mapping/space_management/space_level.dm]: Should this be `Z_LEVEL_NORTH`

File: code/modules/mapping/space_management/zlevel_manager.dm [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		return FALSE
+	var/list/trait_list
+	if(z_list)
+		var/datum/space_level/S = get_zlev(z)
+		trait_list = S.flags
+	else
+		var/list/default_map_traits = DEFAULT_MAP_TRAITS
+		trait_list = default_map_traits[z]
+		trait_list = trait_list[DL_ATTRS]
+	return trait_list[trait] ? TRUE : FALSE
```
Comment (Crazylemon64) [inline code/modules/mapping/space_management/zlevel_manager.dm]: Ah, is this more efficient than `(trait in trait_list)`?

Comment (Crazylemon64) [reply code/modules/mapping/space_management/zlevel_manager.dm]: Also, is this ternary operator here to scrub the contained value to `TRUE` or `FALSE`?
Would it not be better to do `!!trait_list[trait]` instead?

Comment (duncathan) [reply code/modules/mapping/space_management/zlevel_manager.dm]: the ternary operator is more immediately obvious in purpose than double negation is so it's no problem imo

Comment (Crazylemon64) [reply code/modules/mapping/space_management/zlevel_manager.dm]: fair on that point

Comment (RemieRichards) [reply code/modules/mapping/space_management/zlevel_manager.dm]: @Crazylemon64 yes, looking up an associative key (list[key]) is MUCH faster than the ```in``` operator, by a landslide.

File: code/modules/mapping/space_management/heap_space_level.dm (hunk 1 above)
Comment (Cyberboss) [reply code/modules/mapping/space_management/heap_space_level.dm]: Yeah so I noticed, will look into improving the algorithm once i get it working

File: code/modules/mapping/space_management/space_level.dm (hunk 2 above)
Comment (Cyberboss) [reply code/modules/mapping/space_management/space_level.dm]: yes

**Comment by ninjanomnom** [comment]
@Cyberboss This would be a good opportunity to not have *any* static loaded maps. Assign z levels in initialize just based on map configs like a z level group var which determines what is connected where.

You probably want 2 kinds of groups: space transition and multi-z

**Comment by ninjanomnom** [comment]
Also that way we could have centcom with secrets.

**Comment by Cyberboss** [comment]
We need at least centcom so we can put players somewhere when the world boots

**Comment by Cyberboss** [comment]
soon(tm)

**Comment by ninjanomnom** [comment]
A loading screen directly displayed to players if they join before a splash screen is loaded would work to remove centcom.

**Comment by Cyberboss** [comment]
Yeah fuck it lets bite the bullet

**Comment by davethwave** [comment]
Hope this comes to be, and not be swept aside to rot.

**Comment by Cyberboss** [comment]
I'll try and buckle down and do some work on this tonight

On Nov 13, 2017 11:02 AM, "davethwave" <notifications@github.com> wrote:

Hope this comes to be, and not be swept aside to rot.

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/32037#issuecomment-343965630>,
or mute the thread

**Comment by davethwave** [comment]
Thanks, this could improve cpu management, ram, and make navigating space easier and less confusing if I read this right.

**Comment by SpaceManiac** [comment]
Going to be [attempting to pick this up](https://github.com/AutomaticFrenzy/tgstation/commits/space-manager) since I have some future fixes in mind which would really benefit from it, #32701 among them.

**Comment by Cyberboss** [comment]
I'm actually back on this today

File: code/controllers/subsystem/mapping.dm [hunk 4]
```diff
@@ -106,29 +115,38 @@ SUBSYSTEM_DEF(mapping)
...
 	var/start_time = REALTIMEOFDAY
+	INIT_ANNOUNCE("Loading CentCom")
+	TryLoadZ("_maps/map_files/generic/CentCom.dmm", FailedZs)
+
+	//splash turfs are ready, send em in
+	for(var/mob/dead/new_player/NP in GLOB.player_list)
+		NP.SendToStartLocIfRequired()
+
+	INIT_ANNOUNCE("Loading Lavaland")
+	TryLoadZ("_maps/map_files/Mining/Lavaland.dmm", FailedZs)
```
Comment (ninjanomnom) [inline code/controllers/subsystem/mapping.dm]: Shouldn't these be configs?

Comment (Cyberboss) [reply code/controllers/subsystem/mapping.dm]: I'll gladly do it if were ever get a mining level other than lavaland

**Comment by Razharas** [comment]
Did you understand the magic of space transition grid building code or just copied it?
Please be careful its rather tricky

**Comment by Cyberboss** [comment]
I know, I'm being careful with it. Hopefully I can get away with not fully
rewriting it.

**Comment by Razharas** [comment]
You totally can, just use chunks of it, it can be chopped into modules quite well

**Comment by Cyberboss** [comment]
More progress on this coming tonight

**Comment by Razharas** [comment]
I explained how some things work, now its on cyberboss to get into it further and add to the system

**Pull request closed without merging**